### PR TITLE
Enh/special functions

### PR DESCRIPTION
**Pull request by Joel-Dahne**
I have added an interface to [SpecialFunctions.jl](https://github.com/JuliaMath/SpecialFunctions.jl/). All of the code is in `special-functions.jl` where I have written down all the special functions implemented by `SpecialFunction`. For each one I have either given a corresponding Arb implementation or written a small comment about why that one is not handled. Some thinks I though about when doing this:

- How to handle scaled version not implemented by Arb? Should we just make a naive implementation or skip it? If we do implement them the user might assume that they are optimized for the scaled version, on the other hand the user will always get a bound on the error so knows if more precision is needed. If you write generic code with the scaled version it would be nice if it just worked. Maybe we should talk to the developers of `SpecialFunctions` about implementing fallback versions for generic scaled functions?
- Many of the functions take more than one argument, which one should we take the precision from? In most other cases we have taken the maximum of all the inputs so I guess that makes sense here as well? The current implementation doesn't take this into account.
- Should we handle promotion in any way? It would probably be better if this was implemented in `SpecialFunctions` directly.
- Some (which?) Arb functions are more general than those in `SpecialFunctions`. How should we handle that?
- I have not checked that the branch cuts are the same for `SpecialFunctions` and Arb, I believe in most cases they are. 
- Should we add documentation for the methods which are not identical to those in SpecialFunctions (which are these?)? Should we just refer to Arbs documentation?

Do you have any thoughts about these?

### CODE REVIEW

**Comment by kalmarek** [comment]
That is some really meticulous job! My thoughts on the raised points:
* when Arb doesn't implement scaled functions I'd say we should provide fallback; generic fallbacks will not work for Arblib for performance reasons
* I'm fine with the maximum of precisions
* What kind of promotion do you mean? In general I'd stay away from promotion rules as these will only complicate handling precision
* If Arblib functions are more general, I'd leave it to the user to look up the Arblib docs. It's hard to imagine ( at least to me:) ) somebody writing code with low-level Arblib.jl without consulting the original docs. The initial purpose of this package was to provide a user friendly interface to Arb library, so I'd stay away from changing how Arb behaves. Of course we can always discuss the purposes and aims ;), but I'm slightly afraid that producing a user-friendly package for general numerical computations in julia based on Arb needs more time/devotion than I can provide ;)

Would be great to hear @saschatimme take on this.

**Comment by Joel-Dahne** [comment]
Thanks!

- That sounds reasonable I think. I will look at implementing the scaled versions then. 
- Then I'll make sure that all methods use the maximum of the precision.
- For example for `besselj(ν, z)` it's quite common that `ν` is an integer and people might want to call it like `besselj(1::Int, z::Arb)`. But I think we should let this be handled by `SpecialFuntions` and not by us.
- I agree, we can write a note in the documentation referring to the Arb documentation for more details.

I happened to find some time to create this pull request but otherwise I don't have to much time at the moment, so might be a while before I get everything done!

**Comment by Joel-Dahne** [comment]
I did some updates now. More specifically I rewrote the methods to minimize the number of allocations (only one of them allocates things than it returns now) and to use as precision the maximum of the inputs. I also added some methods which I realized I had missed. Finally I added a small section about special functions to the README.

I have not added any scaled versions or versions otherwise not implemented by Arb. It might be something we want to do in the future but I feel like it's not required for this first version.

I'm ready to merge if you are happy with how it looks now!

**Comment by Joel-Dahne** [comment]
I also bumped it to version 0.3.1 now.

**Review by kalmarek — APPROVED** [review]
Looks great ;)

**Pull request merged**
